File: computationalEngineering/SurfboardGeometry/Surfboard/CrossSection.cs
// =============================================================================
// CROSS-SECTION - DECK AND BOTTOM PROFILE AT EACH LONGITUDINAL STATION
// =============================================================================
//
// SURFBOARD CONCEPT: Cross-Sectional Shape
// ==========================================
//
// The cross-section defines the board's shape when sliced perpendicular
// to the length. It varies along the board:
//
//   THICK CENTER (at wide point):      THIN NOSE/TAIL:
//   ─────────────────────────          ──────────────────
//         Deck Crown                     Thin dome
//      ╱‾‾‾‾‾‾‾‾‾‾‾‾╲                  ╱‾‾‾‾‾╲
//     │    Volume!     │               ╱         ╲
//     │                │              │             │
//      ╲___________  ╱                 ╲___________╱
//       Concave                           Flat/Round
//
// CROSS-SECTION ELEMENTS:
// -----------------------
//
// 1. DECK (top surface): Convex dome that adds volume while keeping
//    rails thin. The crown height decreases toward nose and tail.
//
// 2. BOTTOM (bottom surface): Can be flat, single concave, or double
//    concave. Concave channels water flow for speed and lift.
//
// 3. RAILS (edges): The transition from deck to bottom. Ranges from
//    "soft" (round, 50/50 distribution) to "hard" (sharp, down-turned).
//    Hard rails release water cleanly; soft rails are forgiving.
//
//   Rail profiles:
//   ──────────────
//   Soft (round):     Medium (tucked):    Hard (sharp):
//     ╱‾╲               ╱‾╲                 ╱‾╲
//    │   │              │   │               │   │
//    │   │               ╲  │                ╲  │
//     ╲_╱                 ╲_│                 ╲_│
//
// =============================================================================

namespace SurfboardGeometry.Surfboard;

/// <summary>
/// Computes the cross-sectional profile (deck and bottom heights) at a given
/// longitudinal station.
/// </summary>
/// <remarks>
/// For a given position along the board length and across the width,
/// returns the deck height (top) and bottom height relative to the
/// board's center plane. The thickness envelope varies along the board,
/// being thickest at the center and thinning toward nose and tail.
/// </remarks>
public class CrossSection
{
    private readonly SurfboardParameters _params;

    public CrossSection(SurfboardParameters parameters)
    {
        _params = parameters;
    }

    /// <summary>
    /// Get the deck height (top surface Z) at a given longitudinal and lateral position.
    /// </summary>
    /// <param name="t">Normalized longitudinal position (0 = nose, 1 = tail)</param>
    /// <param name="lateralFraction">
    /// Normalized lateral position (0 = centerline, 1 = rail edge).
    /// </param>
    /// <returns>Deck height in mm above the rocker datum (positive = up)</returns>
    /// <remarks>
    /// The deck profile is a cosine dome:
    ///   z_deck = halfThickness(t) + crown(t) * cos(lateralFraction * pi/2)
    ///
    /// The crown (dome height) peaks at the center and falls off toward
    /// the edges (rails), creating a crowned deck that sheds water and
    /// adds volume without increasing rail thickness.
    /// </remarks>
    public float GetDeckHeight(float t, float lateralFraction)
    {
        t = Math.Clamp(t, 0f, 1f);
        lateralFraction = Math.Clamp(lateralFraction, 0f, 1f);

        float thickness = GetLocalThickness(t);
        float halfThick = thickness / 2f;

        // Crown contribution: cosine dome that peaks at centerline
        // and drops to zero at the rail edge
        float crownAtStation = GetLocalCrown(t);
        float crownContribution = crownAtStation * MathF.Cos(lateralFraction * MathF.PI / 2f);

        return halfThick + crownContribution;
    }

    /// <summary>
    /// Get the bottom height (bottom surface Z) at a given longitudinal and lateral position.
    /// </summary>
    /// <param name="t">Normalized longitudinal position (0 = nose, 1 = tail)</param>
    /// <param name="lateralFraction">
    /// Normalized lateral position (0 = centerline, 1 = rail edge).
    /// </param>
    /// <returns>Bottom height in mm below the rocker datum (negative = down)</returns>
    /// <remarks>
    /// The bottom surface includes optional single concave:
    ///   z_bottom = -halfThickness(t) - concave(t) * cos(lateralFraction * pi)
    ///
    /// The concave creates a channel along the centerline:
    ///
    ///   Flat bottom:           Single concave:
    ///   ───────────            ─────╲  ╱─────
    ///                                ╲╱
    ///
    /// The concave depth is greatest in the middle of the board and
    /// fades to zero at the nose and tail.
    /// </remarks>
    public float GetBottomHeight(float t, float lateralFraction)
    {
        t = Math.Clamp(t, 0f, 1f);
        lateralFraction = Math.Clamp(lateralFraction, 0f, 1f);

        float thickness = GetLocalThickness(t);
        float halfThick = thickness / 2f;

        // Concave contribution: cosine valley at the centerline
        // cos(lateralFraction * PI) ranges from 1 (center) to -1 (edge)
        // We want concave at center, so subtract when positive
        float concaveAtStation = GetLocalConcave(t);
        float concaveContribution = concaveAtStation
            * MathF.Pow(MathF.Cos(lateralFraction * MathF.PI / 2f), 2f);

        return -(halfThick + concaveContribution);
    }

    // =========================================================================
    // THICKNESS DISTRIBUTION ALONG THE BOARD
    // =========================================================================

    /// <summary>
    /// Get the local board thickness at a normalized longitudinal position.
    /// </summary>
    /// <remarks>
    /// SURFBOARD: The thickness distribution follows a smooth curve:
    /// - Thin at the nose (~40% of max thickness)
    /// - Maximum at approximately 40% from nose (just ahead of center)
    /// - Gradually thinning toward the tail (~50% of max)
    ///
    /// This uses a cosine-based envelope centered on the thickest point.
    /// </remarks>
    public float GetLocalThickness(float t)
    {
        // Thickest point is typically ~40% from the nose
        float thickestPointT = 0.40f;

        // Minimum thickness fractions at nose and tail
        float noseThicknessFraction = 0.35f;  // 35% of max at nose tip
        float tailThicknessFraction = 0.45f;  // 45% of max at tail tip

        float fraction;
        if (t <= thickestPointT)
        {
            // Nose side: interpolate from nose fraction up to 1.0 at thickest point
            float localT = t / thickestPointT;
            fraction = CosineInterpolate(noseThicknessFraction, 1f, localT);
        }
        else
        {
            // Tail side: interpolate from 1.0 at thickest point down to tail fraction
            float localT = (t - thickestPointT) / (1f - thickestPointT);
            fraction = CosineInterpolate(1f, tailThicknessFraction, localT);
        }

        return _params.MaxThickness * fraction;
    }

    /// <summary>
    /// Get the deck crown height at a normalized longitudinal position.
    /// </summary>
    /// <remarks>
    /// Crown is greatest at the thick center and fades toward nose and tail.
    /// </remarks>
    private float GetLocalCrown(float t)
    {
        // Crown follows a smooth bell curve, peaking at ~45% from nose
        float peakT = 0.45f;
        float distance = MathF.Abs(t - peakT);
        float maxDistance = MathF.Max(peakT, 1f - peakT);
        float normalized = distance / maxDistance;

        // Cosine falloff from peak
        return _params.DeckCrown * MathF.Pow(MathF.Cos(normalized * MathF.PI / 2f), 1.5f);
    }

    /// <summary>
    /// Get the bottom concave depth at a normalized longitudinal position.
    /// </summary>
    /// <remarks>
    /// SURFBOARD: Concave is strongest in the center of the board (under
    /// the surfer's stance) and fades to zero at nose and tail.
    /// The transition zone is the most important section for speed.
    /// </remarks>
    private float GetLocalConcave(float t)
    {
        // Concave is strongest between 30% and 70% of board length
        float startFade = 0.25f;
        float endFade = 0.75f;

        if (t < startFade)
        {
            float localT = t / startFade;
            return _params.BottomConcave * CosineInterpolate(0f, 1f, localT);
        }
        else if (t > endFade)
        {
            float localT = (t - endFade) / (1f - endFade);
            return _params.BottomConcave * CosineInterpolate(1f, 0f, localT);
        }
        else
        {
            return _params.BottomConcave;
        }
    }

    /// <summary>
    /// Cosine interpolation between two values for smooth transitions.
    /// </summary>
    private static float CosineInterpolate(float a, float b, float t)
    {
        float blend = (1f - MathF.Cos(t * MathF.PI)) / 2f;
        return a + (b - a) * blend;
    }
}

File: computationalEngineering/Surfboard/SurfPhysics/geometry/crossSection.py
# -- Surfboard Cross-Section Profile -- #

'''
Computes the cross-sectional profile (deck and bottom heights) at each station.

Direct port of SurfboardGeometry/Surfboard/CrossSection.cs.
Defines the thickness distribution, deck crown dome, and bottom concave
channel at any point on the board surface.

Sean Bowman [02/03/2026]
'''

from __future__ import annotations

import math

from computationalEngineering.SurfPhysics.geometry.parameters import SurfboardParameters


class CrossSection:
    '''
    Surfboard cross-section — deck and bottom heights at any (t, lateralFraction).

    For a given longitudinal position t and lateral position (0 = center, 1 = rail),
    computes the deck (top surface) and bottom surface heights relative to the
    board center plane.

    The thickness envelope varies along the board:
    - Thickest at 40% from nose (100% of maxThickness)
    - Thin at nose (35% of max) and tail (45% of max)
    - Cosine interpolation for smooth transitions
    '''

    def __init__(self, params: SurfboardParameters) -> None:
        '''
        Initialize cross-section from surfboard parameters.

        Parameters:
        -----------
        params : SurfboardParameters
            Board dimensions including crown and concave
        '''
        self._params = params

    def getDeckHeight(self, t: float, lateralFraction: float) -> float:
        '''
        Get the deck height (top surface Z) at a given position.

        The deck profile is a cosine dome that peaks at the centerline
        and drops to zero at the rail edge.

        Parameters:
        -----------
        t : float
            Normalized longitudinal position (0 = nose, 1 = tail)
        lateralFraction : float
            Normalized lateral position (0 = centerline, 1 = rail edge)

        Returns:
        --------
        float : Deck height in mm above the center plane (positive = up)
        '''
        t = max(0.0, min(1.0, t))
        lateralFraction = max(0.0, min(1.0, lateralFraction))

        thickness = self.getLocalThickness(t)
        halfThick = thickness / 2.0

        # Crown: cosine dome peaking at centerline, zero at rail
        crownAtStation = self._getLocalCrown(t)
        crownContribution = crownAtStation * math.cos(lateralFraction * math.pi / 2.0)

        return halfThick + crownContribution

    def getBottomHeight(self, t: float, lateralFraction: float) -> float:
        '''
        Get the bottom height (bottom surface Z) at a given position.

        The bottom surface includes optional single concave — a cosine valley
        at the centerline that fades toward the rails.

        Parameters:
        -----------
        t : float
            Normalized longitudinal position (0 = nose, 1 = tail)
        lateralFraction : float
            Normalized lateral position (0 = centerline, 1 = rail edge)

        Returns:
        --------
        float : Bottom height in mm below center plane (negative = down)
        '''
        t = max(0.0, min(1.0, t))
        lateralFraction = max(0.0, min(1.0, lateralFraction))

        thickness = self.getLocalThickness(t)
        halfThick = thickness / 2.0

        # Concave: cosine^2 valley at centerline
        concaveAtStation = self._getLocalConcave(t)
        concaveContribution = concaveAtStation * math.pow(
            math.cos(lateralFraction * math.pi / 2.0), 2.0
        )

        return -(halfThick + concaveContribution)

    def getLocalThickness(self, t: float) -> float:
        '''
        Get the local board thickness at a normalized longitudinal position.

        Thickness distribution:
        - Thickest at 40% from nose (100% of max)
        - Nose tip: 35% of max
        - Tail tip: 45% of max
        - Cosine interpolation between regions

        Parameters:
        -----------
        t : float
            Normalized position (0 = nose, 1 = tail)

        Returns:
        --------
        float : Thickness in mm
        '''
        # Thickest point at 40% from nose
        thickestPointT = 0.40

        # Min thickness fractions at nose and tail
        noseThicknessFraction = 0.35
        tailThicknessFraction = 0.45

        if t <= thickestPointT:
            # Nose side: interpolate from nose fraction up to 1.0
            localT = t / thickestPointT
            fraction = self._cosineInterpolate(noseThicknessFraction, 1.0, localT)
        else:
            # Tail side: interpolate from 1.0 down to tail fraction
            localT = (t - thickestPointT) / (1.0 - thickestPointT)
            fraction = self._cosineInterpolate(1.0, tailThicknessFraction, localT)

        return self._params.maxThickness * fraction

    def _getLocalCrown(self, t: float) -> float:
        '''
        Get the deck crown height at a normalized longitudinal position.

        Crown is greatest at ~45% from nose and fades toward nose and tail
        using a cosine^1.5 falloff from the peak.

        Parameters:
        -----------
        t : float
            Normalized position (0 = nose, 1 = tail)

        Returns:
        --------
        float : Crown height in mm
        '''
        peakT = 0.45
        distance = abs(t - peakT)
        maxDistance = max(peakT, 1.0 - peakT)
        normalized = distance / maxDistance

        # Cosine falloff from peak
        return self._params.deckCrown * math.pow(
            math.cos(normalized * math.pi / 2.0), 1.5
        )

    def _getLocalConcave(self, t: float) -> float:
        '''
        Get the bottom concave depth at a normalized longitudinal position.

        Concave is at full strength between 25% and 75% of board length,
        with cosine fade-in/fade-out at the ends.

        Parameters:
        -----------
        t : float
            Normalized position (0 = nose, 1 = tail)

        Returns:
        --------
        float : Concave depth in mm
        '''
        startFade = 0.25
        endFade = 0.75

        if t < startFade:
            localT = t / startFade
            return self._params.bottomConcave * self._cosineInterpolate(0.0, 1.0, localT)
        elif t > endFade:
            localT = (t - endFade) / (1.0 - endFade)
            return self._params.bottomConcave * self._cosineInterpolate(1.0, 0.0, localT)
        else:
            return self._params.bottomConcave

    @staticmethod
    def _cosineInterpolate(a: float, b: float, t: float) -> float:
        '''
        Cosine interpolation between two values for smooth transitions.

        Parameters:
        -----------
        a : float
            Start value
        b : float
            End value
        t : float
            Interpolation parameter [0, 1]

        Returns:
        --------
        float : Interpolated value
        '''
        blend = (1.0 - math.cos(t * math.pi)) / 2.0
        return a + (b - a) * blend
